### PR TITLE
fix: added missing text props

### PR DESCRIPTION
**Pull request by pecoram**
fixes https://github.com/lightning-js/renderer/issues/86

Added the following props:

- lineHeight
- maxLines
- verticalAlign
- textBaseline
- overflowSuffix

These props are not yet compatible with the SDF renderer (https://github.com/lightning-js/renderer/issues/107)

More Props not added:
- textOverflow

- [x] visual regression tests @frank-weindel

### CODE REVIEW

File: src/core/text-rendering/renderers/TextRenderer.ts [hunk 1]
```diff
@@ -244,6 +244,10 @@ export interface TrProps extends TrFontProps {
    * @default 0
    */
   letterSpacing: number;
+  lineHeight?: number | null;
+  maxLines?: number | undefined;
+  maxLinesSuffix?: string | undefined;
+  textOverflow?: 'ellipsis' | 'clip' | undefined;
```
Comment (frank-weindel) [inline src/core/text-rendering/renderers/TextRenderer.ts:250]: These shouldn't be optional

File: src/render-drivers/main/MainOnlyTextNode.ts [hunk 2]
```diff
@@ -197,6 +200,46 @@ export class MainOnlyTextNode extends MainOnlyNode implements ITextNode {
     this.coreNode.letterSpacing = value;
   }
 
+  get lineHeight(): ITextNode['lineHeight'] {
```
Comment (frank-weindel) [inline src/render-drivers/main/MainOnlyTextNode.ts:203]: ThreadX support also needed

File: src/core/text-rendering/renderers/TextRenderer.ts [hunk 3]
```diff
@@ -244,6 +244,10 @@ export interface TrProps extends TrFontProps {
    * @default 0
    */
   letterSpacing: number;
+  lineHeight?: number | null;
+  maxLines?: number | undefined;
```
Comment (frank-weindel) [inline src/core/text-rendering/renderers/TextRenderer.ts:248]: Please implement only text `lineHeight` in this PR. `maxLines` is already implemented by setting the height to `lineHeight * numLines` and contain to `both`. There is a plan to handle `maxLinesSuffix` and `textOverflow` differently.

Comment (pecoram) [reply src/core/text-rendering/renderers/TextRenderer.ts]: Are you sure that `maxLines` is not needed? Consider this case:
I want that a node adapts his height based on the text but no more then `lineHeight * numLinesà. So the height can go from `lineHeight` to `lineHeight * maxLines `

File: src/core/text-rendering/renderers/CanvasTextRenderer.ts [hunk 4]
```diff
@@ -317,6 +325,8 @@ export class CanvasTextRenderer extends TextRenderer<CanvasTextRendererState> {
         wordWrapWidth:
           state.props.contain === 'none' ? undefined : state.props.width,
         letterSpacing: state.props.letterSpacing,
+        lineHeight: state.props.lineHeight,
+        maxLines: state.props.maxLines,
```
Comment (frank-weindel) [inline src/core/text-rendering/renderers/CanvasTextRenderer.ts:329]: For a `maxLines` implementation for the CanvasTextRenderer, I'd like to base it off of the `contain` === 'both' mode like so (semi-pseudocode):
```
{
  //...
  maxLines: contain === 'both' ? Math.floor(height / lineHeight) : undefined
}
```
Where `lineHeight` is the set lineHeight property, if provided, or the set `fontSize` property.

SDF Text Renderer already implements this capability in basically this way.

Comment (pecoram) [reply src/core/text-rendering/renderers/CanvasTextRenderer.ts:329]: this is ok but it is not suitable for dynamic height text nodes. See my example image.

Comment (frank-weindel) [reply src/core/text-rendering/renderers/CanvasTextRenderer.ts:337]: @pecoram Where can I find your images?

Comment (pecoram) [reply src/core/text-rendering/renderers/CanvasTextRenderer.ts:337]: https://github.com/lightning-js/renderer/pull/98#discussion_r1424073231

Comment (frank-weindel) [reply src/core/text-rendering/renderers/CanvasTextRenderer.ts:337]: Got it. Yeah that seems like a valid use case to me. We'll just want to make sure that the contained height and maxLines work well together. For instance, if maxLines is 99, but the contained height can only fit 5 lines then only 5 lines are rendered before the text is truncated.

Comment (pecoram) [reply src/core/text-rendering/renderers/CanvasTextRenderer.ts:337]: image

```
height: 36
lineHeight: 36
maxLines: 99
contains: both
```

**Comment by wouterlucas** [comment]
Really cool @pecoram 😎 thanks!

I dont want to be that guy, but doing it anyway, any chance we can get an accompanying examples test that demonstrates/tests this?

Would be nice to have this included in future visual regression tests.

**Comment by pecoram** [comment]
I added some visual regression tests

File: src/core/text-rendering/renderers/LightningTextTextureRenderer.ts [hunk 5]
```diff
@@ -212,7 +212,7 @@ export class LightningTextTextureRenderer {
       this._settings.offsetY === null
         ? null
         : this._settings.offsetY * precision;
-    let lineHeight = (this._settings.lineHeight || fontSize) * precision;
+    let lineHeight = (this._settings.lineHeight || fontSize * 1.4) * precision;
```
Comment (frank-weindel) [inline src/core/text-rendering/renderers/LightningTextTextureRenderer.ts:215]: Where does this `1.4` value come from? One of the reason that many of the existing tests needed to be re-done.

File: src/core/text-rendering/renderers/LightningTextTextureRenderer.ts [hunk 6]
```diff
@@ -673,10 +673,10 @@ export class LightningTextTextureRenderer {
       lineHeight: null,
       textBaseline: 'alphabetic',
       textAlign: 'left',
-      verticalAlign: 'top',
+      verticalAlign: 'middle',
```
Comment (frank-weindel) [inline src/core/text-rendering/renderers/LightningTextTextureRenderer.ts:676]: We should default to verticalAlign `'top'`

**Comment by frank-weindel** [comment]
Everything looks good beside my comments. I went ahead and pushed some new commits that resolve my issues. So I'm ready to merge this now 😄 

Thank you @pecoram !

**Pull request merged**